### PR TITLE
Completed Kudomon Solution

### PR DESCRIPTION
**Pull request by thomasGodfrey**
Finished! ... Hopefully.

### CODE REVIEW

File: kudomonTG/src/main/java/main/KudomonGO.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import exceptions.KudomonCantBeCaughtException;
+import model.ElectricType;
+import model.FireType;
+import model.GameField;
+import model.GrassType;
+import model.Kudomon;
+import model.PsychicType;
+import model.RockType;
+import model.Trainer;
+import model.WaterType;
```
Comment (thomasGodfrey) [inline kudomonTG/src/main/java/main/KudomonGO.java]: That would definitely be more clean!

File: kudomonTG/src/main/java/model/Kudomon.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	private String species;
+	private Position position;
+	private GameField field;
+	private Trainer catcher;
+	
+	private double healthPoints;
+	private double combatPoints;
+	protected HashMap<ElementalType, Double> multiplier = new HashMap<ElementalType, Double>();
+	private Random myTurn;
+	private boolean knockedOut;
```
Comment (thomasGodfrey) [inline kudomonTG/src/main/java/model/Kudomon.java]: My thinking was that most fields are not used by the Kudomon subclasses and so didn't strictly need to be protected. I can understand why having a consistent access modifier would be nicer though

File: kudomonTG/src/main/java/model/Kudomon.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+	/**
+	 * Return the effectiveness multiplier against the enemy Kudomon type
+	 * @param typeIn - Type of the  Kudomon 
+	 * @return double - The amount of damage the attacking Kudomon will apply to the opponent
+	 */
+	public double getEffectiveness(ElementalType typeIn){
+		
+		if(!multiplier.containsKey(typeIn)){
+			return 1.0;
+		}
```
Comment (thomasGodfrey) [inline kudomonTG/src/main/java/model/Kudomon.java]: A default value could be possible... However wouldn't I still be required to check if the HashMap contains the key, to know whether to use the default value or not?

File: kudomonTG/src/main/java/model/Trainer.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		//If the Kudomon is not being caught, then start capture
+		}else if(kudomonIn.isBeingCaughtBy()==null){
+			kudomonIn.setBeingCaughtBy(this);
+			capturingKudomon = kudomonIn;
+		
+		//Otherwise another Trainer has already caught the kudomon, throw an error
+		}else{
+			throw new KudomonCantBeCaughtException(kudomonIn.getSpecies() +" has been stolen from you!");
+		}
+	}
```
Comment (thomasGodfrey) [inline kudomonTG/src/main/java/model/Trainer.java]: Thats a good point, I think I misinterpreted "handle this elegantly"... My elegant solution still makes players sad!
One solution may be to give the first player the original Kudomon and remove it from the Field, but while they are completing the catch, any player that calls 'attemptCapture' will be given a copy of the Kudomon (by editing line 79). Once the first player finishes their catch, no other player will be able to catch the Kudomon as it has still been removed from the gameField.

**Comment by thomasGodfrey** [comment]
Hi Matt, Thank you for the in-depth feedback within such a short amount of time! I really appreciate that. 
I agree that my code style could definitely be more consistent and I will take active steps to improve that in the future. Your other comments, including those on the HashMap and Capture methods, are equally appreciated, I will be attempting an improved solution in my own time based on your feedback.
I found this to be a great task, it was enjoyable and the task domain was fun to work with. The open nature of the task also made certain parts deceptively challenging, which resulted in a rewarding piece of work. 
Thanks again for the great feedback!

File: kudomonTG/src/main/java/model/Kudomon.java (hunk 3 above)
Comment (mattschofield) [inline kudomonTG/src/main/java/model/Kudomon.java]: Good question. A bit of Googling suggests you could override it: http://stackoverflow.com/a/7519422/2056908.

One for future reference, perhaps!

File: kudomonTG/src/main/java/model/Kudomon.java (hunk 2 above)
Comment (mattschofield) [inline kudomonTG/src/main/java/model/Kudomon.java]: I was mainly checking you understood the difference between the two - I think you did the right thing 👍

**Pull request closed without merging**